### PR TITLE
JAMES-3201 Vavr either adoption in reindexing task

### PR DESCRIPTION
**Pull request by LanKhuat**
Related: https://github.com/linagora/james-project/pull/3423

### CODE REVIEW

File: mailbox/tools/indexer/pom.xml [hunk 1]
```diff
@@ -114,6 +114,11 @@
             <artifactId>guice</artifactId>
             <scope>test</scope>
         </dependency>
+        <dependency>
+            <groupId>io.vavr</groupId>
+            <artifactId>vavr</artifactId>
+            <version>0.9.0</version>
```
Comment (blackheaven) [inline mailbox/tools/indexer/pom.xml:120]: specify it in the main pom.xml

**Comment by LanKhuat** [comment]
> where is the XOR?



> where is the XOR?

i pushed a new fixup on https://github.com/linagora/james-project/pull/3423 before squashing all the commit in the NO-REVIEW commit. Did you miss it?

File: mailbox/tools/indexer/src/main/java/org/apache/mailbox/tools/indexer/ReIndexerPerformer.java [hunk 2]
```diff
@@ -278,26 +278,26 @@ public ReIndexerPerformer(MailboxManager mailboxManager,
...
-        return either.either(
-            failure -> {
-                failure.recordFailure(reprocessingContext);
-                return Result.PARTIAL;
-            },
-            result -> result.onComplete(reprocessingContext::recordSuccess));
+    private Result recordIndexingResult(Either<Failure, Result> either, ReprocessingContext reprocessingContext) {
+        return either.fold(failure -> {
+            failure.recordFailure(reprocessingContext);
+            return Result.PARTIAL;
```
Comment (chibenwa) [inline mailbox/tools/indexer/src/main/java/org/apache/mailbox/tools/indexer/ReIndexerPerformer.java:299]: This looks duplicated with reIndex(Either<Failure, ReIndexingEntry> entryOrFailure, ReprocessingContext reprocessingContext) (l 290 -> 293)

File: mailbox/tools/indexer/src/main/java/org/apache/mailbox/tools/indexer/ReIndexerPerformer.java [hunk 3]
```diff
@@ -278,26 +278,26 @@ public ReIndexerPerformer(MailboxManager mailboxManager,
...
-    private Mono<Task.Result> reIndexMessage(Either<Failure, ReIndexingEntry> entryOrFailure, ReprocessingContext reprocessingContext) {
-        return toMono(entryOrFailure.right().map(this::index))
-            .map(this::flapMapRight)
-            .map(either -> recordIndexingResult(reprocessingContext, either));
+    private Mono<Task.Result> reIndex(Either<Failure, ReIndexingEntry> entryOrFailure, ReprocessingContext reprocessingContext) {
+        return entryOrFailure.map(this::index)
+            .fold(failure -> {
+                failure.recordFailure(reprocessingContext);
+                return Mono.just(Result.PARTIAL);
+            }, result -> result.map(either -> recordIndexingResult(either, reprocessingContext)));
```
Comment (chibenwa) [inline mailbox/tools/indexer/src/main/java/org/apache/mailbox/tools/indexer/ReIndexerPerformer.java:293]: As far as I understand it what you can 'result' is a `Mono<Either<Failure, ReIndexingEntry>>` while the overall type is `Either<Failure, Mono<Either<Failure, Result>>>`.

I would expect a trainsformation to go through these steps

 - Move the Mono as the main type: `Mono<Either<Failure, Either<Failure, Result>>>`
 - Then "flatmap" the right part of the expression in order to get a `Mono<Either<Failure, Result>>` wich we would be able to record failures in a single place.

(These transformation are doable as either is a monad, and would IMO lead to a cleaner code style here).

Note that method extraction helps smoothing things (exemple with the previous functionJava version).

File: mailbox/tools/indexer/src/main/java/org/apache/mailbox/tools/indexer/ReIndexerPerformer.java [hunk 4]
```diff
@@ -285,19 +285,19 @@ public ReIndexerPerformer(MailboxManager mailboxManager,
...
 
-    private Mono<Task.Result> reIndex(Either<Failure, ReIndexingEntry> entryOrFailure, ReprocessingContext reprocessingContext) {
-        return entryOrFailure.map(this::index)
-            .fold(failure -> {
-                failure.recordFailure(reprocessingContext);
-                return Mono.just(Result.PARTIAL);
-            }, result -> result.map(either -> recordIndexingResult(either, reprocessingContext)));
+    private Mono<Task.Result> reIndex(Either<Failure, ReIndexingEntry> failureOrEntry, ReprocessingContext reprocessingContext) {
+        return toMono(failureOrEntry.map(this::index))
+            .map(failureOrIndexingResult -> failureOrIndexingResult.getOrElseGet(Either::left))
```
Comment (chibenwa) [inline mailbox/tools/indexer/src/main/java/org/apache/mailbox/tools/indexer/ReIndexerPerformer.java:290]: We could extract this technical step into a "flatMapRight" method toabstract away complexity?

File: mailbox/tools/indexer/src/main/java/org/apache/mailbox/tools/indexer/ReIndexerPerformer.java [hunk 5]
```diff
@@ -309,6 +309,10 @@ private Result recordIndexingResult(Either<Failure, Result> failureOrTaskResult,
             });
     }
 
+    private <X, Y> Either<X, Y> flapMapRight(Either<X, Either<X, Y>> nestedEither) {
```
Comment (remk) [inline mailbox/tools/indexer/src/main/java/org/apache/mailbox/tools/indexer/ReIndexerPerformer.java:312]: s/flatMapRight/flatten ?

**Comment by Arsnael** [comment]
Needs a rebase :)

**Comment by LanKhuat** [comment]
rebased

**Comment by chibenwa** [comment]
```
RabbitMQEventDeadLettersIntegrationTest.groupIdOfFailedEventShouldBeStoredInDeadLetter:320 1 expectation failed.

[594ab53a30c3d7ebd9a35f7ec29fe842f56f5f41] JSON path . doesn't match.
[594ab53a30c3d7ebd9a35f7ec29fe842f56f5f41] Expected: iterable over ["org.apache.james.webadmin.integration.rabbitmq.RabbitMQEventDeadLettersIntegrationTest$RetryEventsListenerGroup"] in any order
[594ab53a30c3d7ebd9a35f7ec29fe842f56f5f41]   Actual: []
```

test this please

**Comment by Arsnael** [comment]
@blackheaven toook your last comment in consideration when merging

Merged.

**Pull request closed without merging**